### ISSUE TITLE
RFE: Support client certificate authentication

### ISSUE THREAD

**Issue by tnozicka** [scylladb/scylladb#10099]
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Currently, scylla supports only password authentication (`authenticator: PasswordAuthenticator`) where the user proves that he is, who he says he is, by presenting a corresponding password. That has several drawbacks:
- Passwords are less secure then certificates
- Can't be setup **before** starting scylla so there is a window where **anyone** can change the password before the real user or create a new *hidden* user
- Management software (like scylla-operator) can't guarantee being able to talk to scylla as it can't guarantee access to cql because the user can change the password at any time
- Management software (like scylla-operator) can't reconcile scylla settings from it's API (like passwords) because it may not have access (changed admin password)

Scylla clients already support mTLS which cryptographically proves that the client talking to the scylla server is who he claims to be. In more detail, scylla server has a configured clientCA (client_encryption_options.truststore) which allows only connections with client certificates signed by this CA. The CA is responsible for verifying the client CommonName (or alternative DNS names) at the time of signing a particular CSR. So when a client presents a client certificate in the SSL flow, scylla can read the CommonName and consider such user as authenticated. (Currently it authenticates again using PasswordAuthenticator.)

Given the users will still go through authorization steps you still need to handle the admin account that can set anything. The pattern used in Kubernetes apiserver is to act on certificates signed for group "system:masters" specially. When client presents a valid certificate with any CN but O="system:masters" he is authorized for any operation. (Also Kubernetes will give system:admin perms only to clients using certificates, not passwords). Groups (Organizations) can also be used for better permission management in general.

Using certificate based authentication would:
- Add more security by using certs instead of passwords
- Remove the need to start with an insecure default password (cassandra/cassandra) that attackers can use to secretly add an extra user before you change it.
- Allow forcing "service/admin users" - e.g. Scylla Operator could manage the ClientCA (accessible to users in the API) and sign it's management cert with it.

Currently, you could use the same cert and still authenticate as multiple users in dedicated connections. [This is handled by impersonation](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation).

Other databases like CockroachDB seem to support cert based auth too https://www.cockroachlabs.com/docs/stable/authentication.html#client-authentication

**What is use case behind this feature:**
- Security
- Manageability

**Additional Information:**
https://kubernetes.io/docs/reference/access-authn-authz/authentication/#x509-client-certs
https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation

**Comment by avikivity** [comment]
@slivne let's prioritize this

**Comment by r0mant** [comment]
Hello folks, is this enhancement on the roadmap? Thanks!

**Comment by Ten0** [comment]
Kafka for that issue provides a simple regex API to map from certificate data to username.

That seems very practical and easy to put in place.

https://docs.confluent.io/5.5.3/security/rbac/configure-mds/mutual-tls-auth-rbac.html#principal-mapping-rules-for-ssl-listeners-extract-a-principal-from-a-certificate

**Comment by nyh** [comment]
I think this is a duplicate of issue #7517.
That issue also refers to a Cassandra patch, https://issues.apache.org/jira/browse/CASSANDRA-14652, which apparently reached Cassandra 4.0 already so we should their choices for compatibility reasons.

**Comment by Ten0** [comment]
I'm not sure that is a duplicate : https://github.com/scylladb/scylladb/issues/7517#issuecomment-723927327

**Comment by nyh** [comment]
@Ten0 both issues sound vague, asking to "accept" (#7517) or "support" (this issue) client certificates, but it's quite obvious to me that the intention of both is use these client certificates to do something useful - namely - authentication replacing password authentication. If there's a difference between these issues, I don't understand what it is.

**Comment by Ten0** [comment]
Well the other one seems unclear enough that it seems  best to close the other one.

**Comment by nyh** [comment]
> Well the other one seems unclear enough that it seems best to close the other one.

Definitely one of the two issues should be closed. Someone who plans to look into this (you or @elcallio I guess) should make that decision. Just try to copy the interesting information that the closed had to the open issue.

> With regards to copying Cassandra, it doesn't seem Cassandra's support following this patch is such that this issue would be resolved, and in that they seem different.

Again, I don't know, and somebody should seriously look into this question. Cassandra apparently added a feature of client certificates in Cassandra 4.0. You'd need to look into exactly what they added, and why, and whether we want to use any of their ideas of APIs or configuration options, etc., in Scylla. If we don't, then users that switch from Cassandra 4.0 to Scylla and already use that new Cassandra 4.0 feature can be disappointed.

**Comment by elcallio** [comment]
@nyh - I don't see the certificates collected for `SaslNegotiator` in cassandra ever being used? Is the actual certificate auth feature restricted to datastax edition?

@tnozicka - For scylla, do we really need anything more complicated than an Authenticator that uses CN=<user/role>? Also, should this be an enterprise feature? (Or even fully internal until evaluated and finalized?)

**Comment by nyh** [comment]
> @nyh - I don't see the certificates collected for `SaslNegotiator` in cassandra ever being used? Is the actual certificate auth feature restricted to datastax edition?

I don't know. The patch was written by Dinesh A. Joshi from Apple, so it's probably not a Datastax-only feature, but perhaps a feature only used by some internal authenticator used by Apple so perhaps, indeed, there is nothing we can learn from that Cassandra feature.

So perhaps we should just close #7517 as a duplicate of this one, if it doesn't contain any interesting information not already in this issue.

**Comment by elcallio** [comment]
@tnozicka it seems minimum config required here is:
* A default superuser role for the auth/role manager 
* One or more mapping expressions to go from full DN to user/role

I am very unsure exactly how the latter should look though. I am not super happy about the kafka example above... At least not until we clearly define how we would this auth to fully work. The OP above suggests that a cert should be allowed to map to a role directly without a user mapping. Not sure about that. It sort of breaks the model a bit... To me it seems nicer if we allowed configuring both super user and admin role names in config, and since no password -> no real security risk. 

The we would be back to mainly just wanting to extract a user from DN which would make any regex or similar easier, since we could just allow an expression on the full DN string.

**Comment by elcallio** [comment]
I would like @tnozicka (and others from serverless) to comment on https://github.com/scylladb/scylladb/pull/12214, which aims to provide the above functionality. Then merge it (or an altered version of it)

**Comment by tnozicka** [comment]
First, apologies for missing the updates here, I had to declare bankruptcy on my GH notifications at some point :(

> To me it seems nicer if we allowed configuring both super user and admin role names in config, and since no password -> no real security risk.

I don't feel strong but having standardized roles with dynamic names will make other things harder.

I think building the rule extraction grammar is gonna be quite painful and I don't see the particular need for it. The user supplies the trusted CA - he thus controls the certs that are generated and can sign them for the particular names.

There is always an option to add it later or accept a community contribution, but I don't see it as MVP.

> A default superuser role for the auth/role manager

yes, a role that allows all actions by default (in my example this would be equivalent to `system:masters`)

To expand on he Kubernetes example, the RBAC validates the x509 Organization (group) against roles and role bindings that define which role (transitively a group) maps to a particular role, like
```
apiVersion: rbac.authorization.k8s.io/v1
# This cluster role binding allows anyone in the "manager" group to read secrets in any namespace.
kind: ClusterRoleBinding
metadata:
  name: read-secrets-global
subjects:
- kind: Group
  name: manager # Name is case sensitive
  apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: ClusterRole
  name: secret-reader
  apiGroup: rbac.authorization.k8s.io
```
To my understanding, in ScyllaDB RBAC model this is simplified as every user maps exactly into one role by name. As roles are hierarchical, a role for a group of users is again a role. With that I assume both `x509.CN` and `x509.O` would map directly to a role in this case. Note that x509 organization is an array. (If we were to accept DNSNames as an equivalent of CN that would be an array too.)

Is is possible to map a user setting both `x509.CN` and `x509.O` to both roles internally? Or in general for multiple x509 organizations (groups)?

**Comment by elcallio** [comment]
@tnozicka If you look at the RFC PR for this, you will see that is does two things: 
1.) Allow defining super user (+ salted pwd) in config to eliminate the initial cassandra:cassandra user/pwd phase
2.) Allow using regex:s to map cert subject (verified) to user/role names. Roles obviously have to be created via cql as usual.

Suggested enhancements are (except for some code cleanups):
1.) Allow multiple mapping expressions
2.) Allow other attributes (i.e. SAN) except for subject attributes. Use the first one matching. 

Would be good if you could weigh in on whether this fulfils your reqs.

**Comment by Ten0** [comment]
> I don't feel strong but having standardized roles with dynamic names will make other things harder.
> 
> I think building the rule extraction grammar is gonna be quite painful and I don't see the particular need for it. The user supplies the trusted CA - he thus controls the certs that are generated and can sign them for the particular names.

If the certificate of the client is determined by something else than its presence in Scylla (e.g, it's the actual domain for its HTTPS interface or something), that seems limiting in that the username would have to be the domain name). Regex seems simple enough. (https://github.com/scylladb/scylladb/issues/10099#issuecomment-1313374617)

**Comment by tnozicka** [comment]
> 1.) Allow defining super user (+ salted pwd) in config to eliminate the initial cassandra:cassandra user/pwd phase

I guess there could be no password / login at all as the x509 authenticator is not using it but I assume it's done so it works with password Auth as well.

Maybe I haven't yet described how this would fall together as I've initially saw it as a next step from the pieces.

I assume we'll need more then one setup, I can think of at least these:
- *use x509 if the role exists or fallback to password*
 Cloud service already use the CN for mTLS but use password auth so for a graceful migration both should work simultaneously. Also if we want to say use cert based auth for super-admin and allow password auth for less priviliged user so say regular apps can share the TLS config / certs and a finer grained access is still managed within ScyllaDB. That will allow us initially to provide a limited set of certs for a few predefined users while still keeping the old ways working by providing a cert with either empty or magic CN that doesn't have a role associated.
- *use x509 only (for security enforcement*)

I think authenticators should form a chain. Say you'd configure
```yaml
authenticators:
- x509Authenticator
- PasswordAuthenticator
```
given this is not an array today, maybe for the better, I suppose we'd have to create a preset for an opinionated DelegatedAuthenticatorX509Passwd which would:
1. [Prereq] verify client cert against client CA (mTLS)
1. [Prereq] if the verification fails, close the connection
1. fetch the `CN` (or regex) and `O` from x509 client cert and match them against a roles
1. if no role matches, move to the next authenticator
1. if more then one role matches use all or a super set if possible

quite similar to (but less complex as we don't have some of the subcases)
https://github.com/kubernetes/kubernetes/blob/322c18c147da08dd2eea25ba3c2b9630a228cf5f/staging/src/k8s.io/apiserver/pkg/authentication/authenticatorfactory/delegating.go#L68-L126


Group wise, it would be nice if we could define roles like `scylladb:admin`, `scylladb:reader` or the customer could define a role like `developers` and when generating a cert for a particular user or app he'd just add a Organization into the x509 to bind the user to the role without having to come back to scylla and have to use CQL to create an aggregated rule for each user / app.

> auth_superuser_name

If I understand that correctly, will this make a dynamically named super user role? Would it be better to create a fixed one that people can *rely on* being present when issuing the certificates and if someone can't sign for that role/CN they could create an "alias" role with CQL for such user and the default role still would be there (In some sense the role name is an API.)

Also one of the points would be that the less we need to have in the config, the less disagreement there can be when doing rolling updates that change such configuration. Say with the regex - what's your upgrade scenario to change it? Say you want to migrate to new cert format (e.g. from pattern A to B or CN to SAN) - how will you make sure that the cert keeps working while rolling out the change and some node have the old value while some have the new? E.g. one of the options would be to define the same Authenticator in a chain with different options but then the config becomes complex so maybe an even more complex regex if even possible.

**Comment by elcallio** [comment]
> > 1.) Allow defining super user (+ salted pwd) in config to eliminate the initial cassandra:cassandra user/pwd phase
> 
> I guess there could be no password / login at all as the x509 authenticator is not using it but I assume it's done so it works with password Auth as well.

So, the config user/pwd is not related to auth cert really, it just plugs a hole + made it easier to test. 

Authentication today is _not_ chained, I am not sure we want it to be either. In general, cert-based auth is much more secure than password based one, and having password fallback seems to just lessen hardening.

I am not wholly against, but doing this would open up several other cans of worms as well, since you could easily create a chain
like "cert, passwd, allowall", which suddenly made everything open. Obviously user error, but it sort of adds another vector to screw up in...

As for rule assignment: Cassandra (and thus scylla) rules state a user must exist. Creating users on-the-fly (or assigning roles on the fly) seems very risky and potentially foot-shootable. Since you have to create certificates anyway, adding a little cql query to whatever scripts does so seems like a relative low hurdle...

You can always create certs that resolve to the same user (shared across customer), but with different serials or whatnot. We can augment auth logs so one can more easily determine which instance of the "developer" role cert was used. 

Fixed super user names (and passwords) are the bane of any security. Having a default seems "stable" enough imho. 

Multiple regex is upcoming. I will also add SAN attribute support to seastar.

**Comment by tnozicka** [comment]
> Authentication today is not chained, I am not sure we want it to be either. In general, cert-based auth is much more secure than password based one, and having password fallback seems to just lessen hardening.

What's your upgrade strategy without it?

> As for rule assignment: Cassandra (and thus scylla) rules state a user must exist. Creating users on-the-fly (or assigning roles on the fly) seems very risky and potentially foot-shootable. Since you have to create certificates anyway, adding a little cql query to whatever scripts does so seems like a relative low hurdle...

This is harder then it sounds to do properly. You need level based logic (not edge triggered) and handle retries for failures which in the end means something needs to periodically query scylla to see if the role is there and with correct settings. As that "script" will be managing houndreds of clusters it can't just pool them so there will likely have to be a small process running alongside scylla to reconcile it, taking extra resources.

Maybe I forgot it from last time, but how do you define the role for the super user? You can't use the CQL because you don't have a cert+role yet, right?

> You can always create certs that resolve to the same user (shared across customer), but with different serials or whatnot. We can augment auth logs so one can more easily determine which instance of the "developer" role cert was used.

I don't think this is how x509 certs are meant to work. Even a single user will have several serial numbers given the cert rotations. If the CN means the group/org then the only component that can identify the user is the issuing CA, as it's the only one that could possibly map SN -> user (although not by standard means, if it's not part of the certificate). x509 certs have common attributes to anyone can identify the user just by common fields and verifying the CA signature.

**Comment by elcallio** [comment]
> > Authentication today is not chained, I am not sure we want it to be either. In general, cert-based auth is much more secure than password based one, and having password fallback seems to just lessen hardening.
> 
> What's your upgrade strategy without it?

Do you mean an upgrade/downgrade of an existing cluster from one auth-mode to another? This is not really anything we have support for today afaik (?), beyond switching backend for handling passwords. While I can see the usefulness in allowing fallback, it is still also complicated + unsupported by the current architecture for auth. I am open to add it I guess, but I would like a PM level decision on it. 

> 
> > As for rule assignment: Cassandra (and thus scylla) rules state a user must exist. Creating users on-the-fly (or assigning roles on the fly) seems very risky and potentially foot-shootable. Since you have to create certificates anyway, adding a little cql query to whatever scripts does so seems like a relative low hurdle...
> 
> This is harder then it sounds to do properly. You need level based logic (not edge triggered) and handle retries for failures which in the end means something needs to periodically query scylla to see if the role is there and with correct settings. As that "script" will be managing houndreds of clusters it can't just pool them so there will likely have to be a small process running alongside scylla to reconcile it, taking extra resources.
> 
??? What would you need anything complicated here? I am assuming a user is per-cluster, and at some point in time said user is created through some blocking procedure. Today, using passwords, you would typically do a CQL query to add user  + password into the cluster. Here you would do the same, minus the password. Then create a certificate authenticating the user. 

> Maybe I forgot it from last time, but how do you define the role for the super user? You can't use the CQL because you don't have a cert+role yet, right?
> 
Super user role is pre-defined (in all active auth modules). You would create a cert signed by your trust auth that says "super user name" in whatever field identifies people, then guard that certificate very well. 

> > You can always create certs that resolve to the same user (shared across customer), but with different serials or whatnot. We can augment auth logs so one can more easily determine which instance of the "developer" role cert was used.
> 
> I don't think this is how x509 certs are meant to work. Even a single user will have several serial numbers given the cert rotations. If the CN means the group/org then the only component that can identify the user is the issuing CA, as it's the only one that could possibly map SN -> user (although not by standard means, if it's not part of the certificate). x509 certs have common attributes to anyone can identify the user just by common fields and verifying the CA signature.

Fields like CN and email can be as non-unique as you like. Again, we are presumably talking about snake-oil certificates created for login purposes, so really, any content except actually IP resolvers etc that actually affect gnutls level auth stuff can be pretty much what you want, as long as the trust verifies it. Again, I don't encourage operating this way, but it most certainly could be done.

**Comment by tnozicka** [comment]
> Do you mean an upgrade/downgrade of an existing cluster from one auth-mode to another? This is not really anything we have support for today afaik (?), beyond switching backend for handling passwords. While I can see the usefulness in allowing fallback, it is still also complicated + unsupported by the current architecture for auth. I am open to add it I guess, but I would like a PM level decision on it.

I agree this may be more of a question for a product manager. I am not sure we can afford to tell the big customers to take full downtime for their clusters if they want to enable cert based auth but I agree it can be a separate step to enable the upgrade path.

> ??? What would you need anything complicated here? I am assuming a user is per-cluster, and at some point in time said user is created through some blocking procedure. Today, using passwords, you would typically do a CQL query to add user + password into the cluster. Here you would do the same, minus the password. Then create a certificate authenticating the user.

- it's all asynchronous + possibly day 2 operations
- there needs to be a loop (program for every scyllacluster with reserved CPU and mem we pay for - the operator can't poll CQL) that retries the create/update calls to reconcile user->role/group binding
- there needs to be a logic handling updates (desired state changes) - requires polling
- there needs to be logic that reconciles the desired state in general - if it were to get deleted by anyone it will recreate the desired state - needs polling

None of this is extremely hard to do, it just felt easier / cheaper to be done in scylla itself. Either way is fine with me.

> Super user role is pre-defined (in all active auth modules). You would create a cert signed by your trust auth that says "super user name" in whatever field identifies people, then guard that certificate very well.

good, thx.

> Fields like CN and email can be as non-unique as you like. 

Have users Alice and Bob. To share a role/group you can't issue distinct certs `CN=alice,O=developer` and `CN=bob,O=developer` to share a developer role. Instead you'd have to issue `CN=developer,MAIL=alice@scylladb.com` and `CN=developer,MAIL=bob@scylladb.com`. But you are right that since the reqex can map only the CN to a role there is still place to put some additional info that helps to identify the real user instead of the CN. Putting the group into the CommonName field make the semantics to be not as pretty but that's something I can live with.

(My idea was that `CN=alice,O=developer` would give you a both roles (super set) for `alice` and `developer`. `CN=developer,MAIL=alice@scylladb.com` can't semantically achieve that.)

**Comment by Ten0** [comment]
> to share a developer role [...] Putting the group into the CommonName field make the semantics to be not as pretty  

It looks like thanks to the array-of-objects design of the settings, this could easily be solved by adding a `role` key that would override extracting a role from the regex and instead just use the provided role (but still *validate* using the regex). 

(this specific evolution being simple to add as the need arises was originally mentioned as one of the benefits of the array-of-object design for the settings 🙂 - https://github.com/scylladb/scylladb/pull/12214#discussion_r1162453487)

**Comment by elcallio** [comment]
> > 
> 
> Have users Alice and Bob. To share a role/group you can't issue distinct certs `CN=alice,O=developer` and `CN=bob,O=developer` to share a developer role. Instead you'd have to issue `CN=developer,MAIL=alice@scylladb.com` and `CN=developer,MAIL=bob@scylladb.com`. But you are right that since the reqex can map only the CN to a role there is still place to put some additional info that helps to identify the real user instead of the CN. Putting the group into the CommonName field make the semantics to be not as pretty but that's something I can live with.
> 
> (My idea was that `CN=alice,O=developer` would give you a both roles (super set) for `alice` (user) and `developer` (group). `CN=developer,MAIL=alice@scylladb.com` can't semantically achieve that.)

I don't think that is what you want. Remember that "users" are just another role. You create a "group" role, then you create a user role and add it as a member of the group. The authenticator should _never_ create any data nor any relation between them, only allow resolving defined relations. 

While you could resort to just using the group as a principal and have the "user" be an ephemeral part of the certificate used, I think that is a bad idea. In my mind, this should just be a means of avoiding passwords, having a more solid authenticator, not an alternative way of "creating" roles/users/etc or simulate them. All the basic create roles+add relations still need to happen via normal CQL queries, pre-existing to login. Only password distribution is bypassed.
